[BEAM-1960] Add Kubernetes scripts for clusters for Performance and Integration tests of Cassandra and ES for Hadoop Input Format IO

diptikul opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

diptikul commented:
Added Loadbalancer, will add the commit. For the pre built image https://github.com/k8s-for-greeks/docker-cassandra-k8s, this has exactly same docker file as we referred others. In the docker file, there  is a check if the image is being built for production usage, there is a script for removal of a folder which contains cqlsh, nodetool etc, only if the image is being built for  development, then the removal does not happen. All the published images on quay are for production, and hence we cannot do a cqlsh for data load and show-health to contain nodetool. in our custom built image we have removed this logic of checking for production or development, hence it works seamlessly on our custom image.

ssisk commented:
@diptikul Now that PR 2193 is merged, let's make sure this PR gets rebased onto those changes - basically, we should see all the old commits from the former PRs disappear from this PR.

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/teardown.sh:
```diff
@@ -0,0 +1,24 @@
+# Delete Cassandra services and statefulset.
+kubectl delete -f cassandra-service.yaml
+kubectl delete -f cassandra-external-service.yaml
+kubectl delete -f cassandra-statefulset.yaml
```
when I'm running teardown and then setup again, I see the error:
Failed to attach volume "..." on node "..." with: googleapi: Error 400: The disk resource 'projects/.../zones/us-central1-a/disks/...' is already being used by 'projects/.../zones/us-central1-a/instances/...

I earlier asked you to remove the disks and switch to normal disks, so that might resolve this issue, but if not,

either way, make sure you can do a full setup -> teardown -> setup cycle

diptikul replied:
Since we changed storage class to 'anything', and also added deletion of pvc (persistent volume)  in the script of teardown, this problem is resolved.

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-statefulset.yaml:
```diff
@@ -0,0 +1,113 @@
+    spec:
+      containers:
+      - name: cassandra
+        image: us.gcr.io/sincere-nirvana-150207/docker-cassandra
```
also - that image is not public, so even if I wanted to try and use it, I can't

diptikul replied:
Hi Stephen,

we have successfully used Quay image as well as docker hub Cassandra image for using remote cqlsh. We have updated the scripts for start, stop, data-load.
Basically we did point (1) "1.find a second docker image that contains those tools and have them connect to the cassandra cluster in the other kubernetes pods. You can even load it into the k8s cluster as a singleton pod at the beginning of data-load.sh, and then remove it from the cluster"
Removed custom docker image and its related files

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-statefulset.yaml:
```diff
@@ -0,0 +1,113 @@
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 1Gi
```
we probably need more room than 1Gi?

diptikul replied:
changed it to 10 Gi, as it is per pod storage, total consumption will be 30 Gi

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-statefulset.yaml:
```diff
@@ -0,0 +1,113 @@
+        requests:
+          storage: 1Gi
+---
+kind: StorageClass
```
Got your point, ok changed to "anything".  Removed storage-class which was defined earlier.

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-service.yaml:
```diff
@@ -0,0 +1,28 @@
+#    Licensed to the Apache Software Foundation (ASF) under one or more
```
Done- combined Cassandra-service and Cassandra-statefulset and renamed Cassandra-external-service to cassandra-service-for-local-dev

ssisk commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -8,6 +8,6 @@ quickly and easily:
  - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
```
hey dipti - it looks like in this commit you did a merge instead of a rebase - we can't accept PRs that have old commits by other people in them, so we can't use this PR as-is.

I'm not sure if it's easier for you to:
1. use git cherry-pick to pull the commits from this branch into a new branch (and potentially start a new PR?)
2. use git reset to rewind back to before all these changes and then re-apply your changes 
(or something else I haven't thought of :)

regardless, we can't have these changes in this PR

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -84,3 +86,7 @@ echo "-----------------------------"
+
+# Delete cassandra sinle node set up
+cd ../LargeITCluster
+kubectl delete -f cassandra-svc-temp.yaml
```
One solution is that it may be appropriate for this script to not fail on the first error

ssisk replied:
oh! And perhaps this might be the right solution: http://redsymbol.net/articles/bash-exit-traps/

diptikul replied:
Thanks for the solution :) added as per the reference

ssisk commented:
R @ssisk

ssisk commented on sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOTest.java:
```diff
@@ -472,6 +472,7 @@ public void testReadWithNullCreateRecordReader() throws Exception {
    */
```
it seems like we're changing this file in both PRs - we should pick one, otherwise there'll be a merge conflict

diptikul replied:
ok, I think will check into the other PR , will remove this commit

diptikul replied:
Cleaned his commit and fixed this on PR 2460 #2460

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -87,6 +116,5 @@ echo "-----------------------------"
-cd ../LargeITCluster
-kubectl delete -f cassandra-svc-temp.yaml
+# Delete cassandra single node set up before exit 
+trap delete_service EXIT
```
I think you might have to put this up at the beginning of the file (maybe just after the function is defined?), otherwise I don't know that it works

diptikul replied:
Checking and changing

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -20,6 +20,27 @@
+service_list=( $services_exists )
+
+function delete_service {
+cd ../LargeITCluster
```
indent?

ssisk replied:
what I mean by this is: please indent a little if that's the norm for bash scripts

diptikul replied:
Sure will check the norm and make changes accordingly

diptikul commented on sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOTest.java:
```diff
@@ -546,6 +546,34 @@ public void testReadersGetFractionConsumed() throws Exception {
+    boundedSource.setInputFormatObj(mockInputFormat);
+    BoundedReader<KV<Text, Employee>> boundedReader = boundedSource.createReader(p.getOptions());
+    assertEquals(true, boundedReader.start());
+    assertEquals(null, boundedReader.getFractionConsumed());
```
We added

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -0,0 +1,120 @@
+echo "-----------------------------"
+# Record count set to 1000, change this value to load as per requirement.
+# dataintegrity flag is set to true to load deterministic data
+./bin/ycsb load cassandra-cql -p hosts=${external_ip} -p dataintegrity=true -p recordcount=\
```
you mentioned adding a -threads parameter - let's do that here

diptikul replied:
Yes, adding

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-svc-statefulset.yaml:
```diff
@@ -0,0 +1,120 @@
+      accessModes: [ "ReadWriteOnce" ]
+      resources:
+        requests:
+          storage: 10Gi
```
let's make sure this is enough - have you run it with the 50million records data size?

diptikul replied:
Loading data with 50 m, lets see how it goes

diptikul replied:
This is sufficient for 50m

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/elasticsearch/SmallITCluster/elasticsearch-svc-rc.yaml:
```diff
@@ -47,6 +47,18 @@ spec:
     metadata:
```
I realized this is still set up as a public LoadBalancer kubernetes service - we should switch it NodePort and add the separate elasticsearch-service-for-local-dev.yaml  that exposes a public LoadBalancer service. 

Same for the large elasticsearch cluster if that hasn't been similarly switched.

diptikul replied:
Will do the needful

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -20,8 +20,9 @@
 
```
Removed

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-svc-statefulset.yaml:
```diff
@@ -27,7 +42,7 @@ spec:
       containers:
       - name: cassandra
-        image: us.gcr.io/sincere-nirvana-150207/docker-cassandra
+        image: quay.io/vorstella/cassandra-k8s:latest
```
Added a tag v1.2 which contains Cassandra 3.10 installation

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -42,6 +43,8 @@ do
    i=$((i+1))
 done
 
+temp_running_seed="$(kubectl get pods -l name=cassandra-temp -o jsonpath="{.items[0].metadata.name}")"
```
Fixed, added while loop to wait until the pod is in ready state

diptikul commented:
@ssisk , can you please have a look at todays commits ? I have checked in external service for Cassandra small cluster, and changed the query test as it was failing due to read timeout, also added "create index" on a column used in query in the data load script

ssisk commented:
You'll need to merge in the changes that got checked in today in order to resolve the merge conflicts

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -0,0 +1,131 @@
+
+for i in "${service_list[@]}"
+do
+   if [ "$i" == "services/cassandra-temp" ]; then
```
does this work? kubectl get svc -o=name | grep cassandra-temp
then I think you could just do an if statement and skip the whole list iteration thing

ssisk replied:
I bet there's also a way to use "-l X" as is done on line 65, but to be honest, I'd rather just finish this PR 

(both these suggestions are non-blocking the PR - I'd rather not take a risky change to this and get it done by tomorrow)

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/teardown.sh:
```diff
@@ -0,0 +1,25 @@
+kubectl delete -f cassandra-svc-statefulset.yaml
+kubectl delete -f cassandra-service-for-local-dev.yaml
+# Delete the persistent storage media for the PersistentVolumes
+kubectl delete pvc -l app=cassandra
```
well that's important. :) thanks for catching that.

ssisk commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/SmallITCluster/data-load.sh:
```diff
@@ -14,26 +14,41 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# Hashcode for 1000 records is 5ea121d90d95c84076f7556605080f4b2c3081a7, 
```
update hashes please

ssisk commented:
also - this PR needs a jira issue filed for it :)

ssisk commented:
... and it needs the PR title changed to include the jira issue's number

diptikul commented:
Created JIRA ticket https://issues.apache.org/jira/browse/BEAM-1960, adding it in title now.

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/data-load.sh:
```diff
@@ -0,0 +1,131 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Hashcode for 50m records is c89f996324e4dd4b6071a9b735281e0329a58b75
```
Updated

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-svc-statefulset.yaml:
```diff
@@ -0,0 +1,113 @@
+  ports:
+    - port: 9042
+  selector:
+    app: cassandra
```
Changed

diptikul commented on sdks/java/io/hadoop/jdk1.8-tests/src/test/resources/kubernetes/cassandra/LargeITCluster/cassandra-svc-statefulset.yaml:
```diff
@@ -0,0 +1,113 @@
+            exec:
+              command: ["/bin/sh", "-c", "PID=$(pidof java) && kill $PID && while ps -p $PID > /dev/null; do sleep 1; done"]
+        env:
+          - name: MAX_HEAP_SIZE
```
Without MAX_HEAP_SIZE or HEAP_NEWSIZE, our test failed with the timeout. We tested it by removing

ssisk commented on .test-infra/kubernetes/cassandra/LargeITCluster/start-up.sh:
```diff
@@ -17,13 +17,6 @@
-tar xfz slf4j-1.7.22.tar.gz
-cp slf4j-1.7.22/slf4j-simple-*.jar ycsb-0.12.0/lib/
-cp slf4j-1.7.22/slf4j-api-*.jar ycsb-0.12.0/lib/
-echo "YCSB tool loaded"
```
looks like this file got lost in the renames? Should not block merge today, we can add it back in

ssisk commented:
retest this please

dhalperi commented:
Merging based on @ssisk 's LGTM and manual testing.

Expectation is we'll have automated testing someday and better documentation that would set the rest of the community up to be able to understand and review these things :), but this is a nice set of things to start off the process. Thanks @diptikul and @ssisk
